### ISSUE TITLE
markdown class is broken

### ISSUE THREAD

**Issue by Omikhleia** [sile-typesetter/sile#1336]
SILE 0.12.2

_First feeling,_ the **markdown** class is nowhere documented properly, this might be bad sign...

Input document is a random markdown sample taken [here](http://www.unexpected-vortices.com/sw/rippledoc/quick-markdown-example.html).

Preamble is:
```
\begin[papersize=7.44in x 9.68in, class=markdown]{document}
\language[main=en]
\font[family=Libertinus Serif, size=10pt]

\end{document}
```

So going for: `sile  -I markdown-preamble.sil markdown.md` 

I made an early attempt before and got an error. I didn't record it at the time, and decided to give it another fresh try.
```
Error detected:
	./lua-libraries/lunamark/reader/markdown.lua:23: attempt to call field 'lower' (a nil value)
```
That is different that what I remembered, but along the same lines (UTF8 related)... Then I realized I now have on my computer an "utf8" package in some external dependency which I hadn't before... So SILE's bundled lunamark is requiring an utf8 package it doesn't provide...

I know we have "lua-utf8", though... Out of curiosity, I Hacked two occurrence of `require("utf8")` to `require("lua-utf8")` in the lunamark code... And got a document... Yeah... Ah wait, it's a joke. Paragraphs and text contents are _mostly_ there, but without any formatting...

![image](https://user-images.githubusercontent.com/18075640/156878652-4a19f874-73ee-4854-a6a8-a853343fb9cb.png)

What have I done wrong? How is this supported to work.

_Second feeling,_ this is very probably highly broken... Get me wrong. But if this is so broken, why is it here in the classes folder, without any caution? Oh well (take this with a grain of salt and humor), I _already_ know that all default classes are broken, the book class in the first place. Why should I be surprised ? :clown_face:

**Comment by Omikhleia** [comment]
Second thought: if the issues are confirmed (and a mistake on my side), then move this class in something such as `classes/unsupported` or whatever. Don't let it live with supposedly working classes until it's not production-ready and feature-full.

**Comment by alerque** [comment]
I do kind of agree. This and the `docbook` class (#1338) are both basically *Proof of Concept* stage classes and are better suited to be examples than packaged and distributed implementations. At least that's my first gut level reaction.

I do a *ton* of work typesetting Markdown but almost never use this class. In fact I never use the class at all, although I do occasionally use some of the related utility functions like `SILE.processMarkdown()` for very short strings that just have a little bit of markup (such as meta data comming from a YAML file). For book content I always use my Pandoc fork to convert Markdown to SILE format and then typeset it directly.

One of the major reasons for that is the Pandoc Markdown parsing is just so much more robust and complete than what we bundled for the Markdown class. At the very least if this is going to change from a POC to a fully supported class we should use a better parser (there are good options).

**Comment by Omikhleia** [comment]
I have made an attempt at fixing this - and while I am getting somewhere more decent than the above attempt, **lunamark** needs an heavy lifting to support many of the Pandoc-extended constructs that would be needed for any good output, e.g.
- Super/sub-scripts, struck-out text...
- Class-like annotations (such as `{-}` or `{.unnumbered}`)
- And plenty of other things... Esp. fenced line blocks and tables...

I have started frantically hacking **lunamark** on my local copy, but this beast is hard to tame. Is this something the direction one should take? That library has not seen any release for years now.

**Comment by alerque** [comment]
No, I suggest ditching Lunamark and using something else to convert Markdown to an AST.

Off the top of my head [cmark-lua](https://github.com/jgm/cmark-lua) is probably the most likely candidate. I started work on a pure Lua [commonmark-lua](https://github.com/alerque/commonmark-lua) a while back but didn't get particularly far. Writing our own Lua wrapper for [comrak](https://github.com/kivikakk/comrak) is also not out of the question. Of course calling out to Pandoc and just getting the AST (native or json or hook in as a Lua writing) is another potential avenue, but that has a major disadvantage of a 3rd party dependency at run time that does not package well inside the Lua ecosystem. I would rather have a markdown inputter that handles basic cases using things that LuaRocks can install natively, then having a separate Pandoc alternative processor.

**Comment by Omikhleia** [comment]
Unless mistaken, my problem is that `cmark` and friends seem to implement CommonMark **only**, that is without many Pandoc extensions (fenced divs with `:::`, line blocks with `|`, raw inlines with `{=xxx}`, attributes such as `{.unnumbered custom-style=MyPara}` on many blocks such as sections, images, code, etc.) -- So it falls short in terms of features compared with what I can achieve with a [Pandoc custom Lua writer](https://github.com/Omikhleia/omikhleia-sile-packages/blob/main/examples/markdown-sample.pdf). In other terms, this is not something I would use, except for very simple cases, as it would lack many advanced features of the kind I would expect for book writing... Where one may have many use cases for it (e.g. `[Ceci est en français]{lang=fr}`...)

**Comment by alerque** [comment]
I agree, there is some utility to having a basic Markdown parser within reach, but for serious use I use Pandoc extensions as well — in fact I've been using Pandoc→SILE exclusively for book production for 6 years now. All the things you mention such as fenced blocks, line blocks, raw inlines, and language attributes are my bread and butter.

That being said implementing a Pandoc compliant reader from scratch is not a trivial project and I'm not interested really—there are two many edge cases. Common mark is far easier to implement and much more reasonable to target as initial support.

Lets let Pandoc parse Pandoc flavored Markdown and pass it to us (whether via a Haskell writer as I use or a Lua writer as you've been working on or as a raw AST like we could write a SILE inputter for).

**Comment by Omikhleia** [comment]
Let's split the difference (or, as we say in French, not sure it translates well, but you might enjoy the idiomatic "let's cut the pear in half").
- There would still be a way to make the lunamark-based "native route" minimally working... rather than staying with the ugly and disappointing output I got above and/or just ditching it...
- There might even be a few doable improvements beyond that. Nope, I am not going to tackle with table parsing in LPEG... But some simpler Pandoc extensions are within reach...

I've opened a draft PR branch where I'll push my proposals, fixes and improvements. Eventually, this might be better anyway than the mess we currently have.
